zerver/migrations/0436_fix_deleteduser_email.py

```diff
@@ -0,0 +1,66 @@
+# Generated by Django 4.2 on 2023-04-17 18:25
+
+from email.headerregistry import Address
+from functools import lru_cache
+
+from django.conf import settings
+from django.core.exceptions import ValidationError
+from django.core.validators import validate_email
+from django.db import migrations
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from django.db.models import Q
+
+
+@lru_cache
+def get_fake_email_domain(apps: StateApps, realm_id: int) -> str:
+    Realm = apps.get_model("zerver", "Realm")
+    realm = Realm.objects.get(id=realm_id)
+
+    # Build out realm.host
+    subdomain = realm.string_id
+    if subdomain == "":
+        host = settings.EXTERNAL_HOST
+    else:
+        default_host = f"{subdomain}.{settings.EXTERNAL_HOST}"
+        host = settings.REALM_HOSTS.get(subdomain, default_host)
+
+    # Implement get_fake_email_domain
+    try:
+        validate_email(Address(username="bot", domain=host).addr_spec)
+        return realm.host
+    except ValidationError:
+        pass
+
+    try:
+        # Check that the fake email domain can be used to form valid email addresses.
+        validate_email(Address(username="bot", domain=settings.FAKE_EMAIL_DOMAIN).addr_spec)
+        return settings.FAKE_EMAIL_DOMAIN
+    except ValidationError:
+        raise Exception(
+            settings.FAKE_EMAIL_DOMAIN + " is not a valid domain. "
+            "Consider setting the FAKE_EMAIL_DOMAIN setting."
+        )
+
+
+def fix_invalid_emails(apps: StateApps, schema_editor: BaseDatabaseSchemaEditor) -> None:
+    UserProfile = apps.get_model("zerver", "UserProfile")
+    invalid_users = UserProfile.objects.filter(is_active=False).filter(
+        Q(email__icontains="@https://") | Q(email__icontains="@http://")
+    )
+    for invalid_user in invalid_users:
+        local_part = invalid_user.email.split("@")[0]
+        invalid_user.email = local_part + "@" + get_fake_email_domain(apps, invalid_user.realm_id)
+        invalid_user.save(update_fields=["email"])
```

Comment:
It's a little sad we need to do parsing via `split("@")`, but I checked the `Address` library can't actually parse them properly, presumably because of how they're invalid.

Reply:
Yes, that failure to parse is the whole reason we have to do this fix. :)